storage/operation/transaction_results.go

```diff
@@ -9,11 +9,50 @@ import (
 	"github.com/onflow/flow-go/storage"
 )
 
-func InsertTransactionResult(w storage.Writer, blockID flow.Identifier, transactionResult *flow.TransactionResult) error {
+// InsertAndIndexTransactionResults inserts and indexes multiple transaction results in a single batch write.
+// The caller must hold the [storage.LockInsertAndIndexTxResult] lock.
+// It returns [storage.ErrAlreadyExists] if transaction results for the block already exist.
+func InsertAndIndexTransactionResults(lctx lockctx.Proof, rw storage.ReaderBatchWriter, blockID flow.Identifier, transactionResults []flow.TransactionResult) error {
+	if !lctx.HoldsLock(storage.LockInsertAndIndexTxResult) {
+		return fmt.Errorf("InsertTransactionResult requires LockInsertAndIndexTxResult to be held")
+	}
+
+	// Check if transaction results for the block already exist
+	// We can exit early if we find one existing transaction result R, assuming that the process which wrote R in the past
+	// correctly inserted all other results for the block containing R.
+	// This function only inserts new transaction results; it does not sanity check existing results or ever overwrite results.
+
+	prefix := MakePrefix(codeTransactionResult, blockID)
+	checkExists := func(key []byte) error {
+		return fmt.Errorf("transaction results for block %v already exist: %w", blockID, storage.ErrAlreadyExists)
+	}
+	err := IterateKeysByPrefixRange(rw.GlobalReader(), prefix, prefix, checkExists)
+	if err != nil {
+		return err
+	}
+
+	// there is no existing transaction result for the block, we can proceed to insert
```

Comment:
What happens for an empty block? I would suggest to add a clarifying comment to the implementation - even if that explanation happens to be "for the empty block, we cannot currently prevent erroneous writes persisting transaction results".